Server/JokesOnYou.Web.Api/Repositories/JokesRepository.cs

```diff
@@ -35,5 +35,40 @@ public async Task<IEnumerable<JokeReplyDto>> GetAllJokeDtosAsync()
 
             return jokeDtos;
         }
+
+        public async Task<Joke> GetJokeToUpdate(JokeUpdateDTO jokeUpdateDTO)
+        {
+            var jokeToUpdate = await GetJokeById(jokeUpdateDTO.Id);
+            
+            //var jokeUpdateDto = await _context.Jokes.ProjectTo<JokeUpdateDTO>(_mapper.ConfigurationProvider)
+            //    .FirstOrDefaultAsync(x => x.Id == jokeUpdateDTO.Id);
+
+            return jokeToUpdate;
+        }
+
+        public async Task Save()
+        {
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task<JokeReplyDto> GetJokeReplyById(int id)
+        {
+            var joke = await GetJokeById(id); 
+
+            if (joke == null)
+            {
+                throw new NotImplementedException("can't find joke"); 
+            }
+
+            var jokeReplyDto = _mapper.Map<Joke, JokeReplyDto>(joke);
+            return jokeReplyDto;
+        }
+
+        public async Task<Joke> GetJokeById(int id)
+        {
+            var joke = await _context.Jokes.FirstOrDefaultAsync(x => x.Id == id);
```

Comment:
@chrisK00, I saw this in other places as well. But shouldn't it be SingleOrDefault? Because we only expect the ID to exist once...?

Reply:
SingleOrDefault will go over the whole table no matter if its already found the object. it will continue checking if there is another object that satisfies the Predicate if for whatever reason two are found it will return null.
FirstOrDefault will stop when 1 is found.

if I did not get this correct please let me know.

Reply:
SingleOrDefault will return an Exception if more than one object is found, not null. It will only return null when nothing is found. 

I prefer this method over FirstOrDefault because it ensures that only one item is found and otherwise let's you know via a Exception that something is wrong. However, FirstOrDefault will indeed be faster. While both are still N(1) I think.

I guess this is comes down to taste, I personally like the explicitly of saying: I want this one thing and I expect that there is exactly one thing that matches my search.

Reply:
I use singleordefault when im searching by name for example because there is a chance we have duplicates since its us developers who did it, but when searching for ids i use firstordefault because our db should auto increment and not allow duplicates